service/history/api/startworkflow/api.go

```diff
@@ -41,167 +46,463 @@ import (
-					runID.String(),
-					startRequest,
-					nil)
-				if err != nil {
-					return nil, nil, err
-				}
-				return workflowContext.GetContext(), workflowContext.GetMutableState(), nil
-			},
-			shard,
-			workflowConsistencyChecker,
+				runID,
+				s.request,
+				nil)
+			if err != nil {
+				return nil, nil, err
+			}
+			mutableState := workflowContext.GetMutableState()
+			mutableStateInfo, err = extractMutableStateInfo(mutableState)
+			if err != nil {
+				return nil, nil, err
+			}
+			return workflowContext.GetContext(), mutableState, nil
+		},
+		s.shardCtx,
+		s.workflowConsistencyChecker,
+	)
+	return mutableStateInfo, err
+}
+
+// respondToRetriedRequest provides a response in case a start request is retried.
+func (s *Starter) respondToRetriedRequest(
+	ctx context.Context,
+	runID string,
+) (*historyservice.StartWorkflowExecutionResponse, error) {
+	request := s.request.StartRequest
+
+	if !request.GetRequestEagerExecution() {
+		return &historyservice.StartWorkflowExecutionResponse{
+			RunId: runID,
+		}, nil
+	}
+
+	// For eager workflow execution, we need to get the task info and history events in order to construct a poll response.
+	mutableStateInfo, err := s.getMutableStateInfo(ctx, runID)
+	if err != nil {
+		return nil, err
+	}
+
+	// The current workflow task is not inflight or not the first task or we exceeded the first attempt and fell back to
+	// matching based dispatch.
+	if !mutableStateInfo.hasInflight || mutableStateInfo.workflowTaskInfo.StartedEventID != 3 || mutableStateInfo.workflowTaskInfo.Attempt > 1 {
+		return nil, serviceerror.NewWorkflowExecutionAlreadyStarted(
+			"First workflow task is no longer inflight, retried request came in too late",
+			request.RequestId,
+			runID,
 		)
-		switch err {
-		case nil:
-			return &historyservice.StartWorkflowExecutionResponse{
-				RunId: runID.String(),
-			}, nil
-		case consts.ErrWorkflowCompleted:
-			// previous workflow already closed
-			// fallthough to the logic for only creating the new workflow below
-		default:
+	}
+
+	events, err := s.getWorkflowHistory(ctx, mutableStateInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.generateResponse(runID, mutableStateInfo.workflowTaskInfo, events)
+}
+
+// getMutableStateInfo gets the relevant mutable state information while getting the state for the given run from the
+// workflow cache and managing the cache lease.
+func (s *Starter) getMutableStateInfo(ctx context.Context, runID string) (*mutableStateInfo, error) {
+	// We techincally never want to create a new execution but in practice this should not happen.
+	workflowContext, releaseFn, err := s.workflowConsistencyChecker.GetWorkflowCache().GetOrCreateWorkflowExecution(
+		ctx,
+		s.namespace.ID(),
+		common.WorkflowExecution{WorkflowId: s.request.StartRequest.WorkflowId, RunId: runID},
+		workflow.CallerTypeAPI,
+	)
+	var releaseErr error
+	defer func() {
+		releaseFn(releaseErr)
+	}()
+
+	if err != nil {
+		return nil, err
+	}
```

Comment:
if err from GetOrCreateWorkflowExecution is not nil, then releaseFn is nil. so defer should be after this check.